### PR TITLE
2914 show last login date

### LINKED ISSUE #2914: Feature Request: Show last login (use) date of a user.
I have a couple of users in my instance. It would be nice to identify inactive users for a cleanup.

### PR DESCRIPTION
**Pull request by equinoxmatt**
Closes #2914

Changes proposed in this pull request:

- Upon login, user configuration file is 'touched'
- Displays last login date on manage users and user detail screen

How to test the feature manually:

1. Create user.
2. Confirm 'Last Login' column on manage users screen and detail screen is 'Never Logged In'
3. Login as new user
4. Confirm 'Last Login' displays correct timesamp.

Pull request checklist:

- [x] clear commit messages
- [x] code manually tested
- [] unit tests written (optional if too hard)
- [ ] documentation updated

[Additional information can be found in the documentation](https://github.com/FreshRSS/FreshRSS/tree/master/docs/en/developers/04_Pull_requests.md).

I was wondering about time formatting. I have stuck with the basic ISO 8601 format, but it isn't the most user friendly display. How are dates handled generally within the project?

Screenshot from 2020-04-30 10-33-28

Screenshot from 2020-04-30 10-33-52

### CODE REVIEW

**Comment by Alkarex** [comment]
Not a strong opinion, but maybe instead of login we could say a more general "last user activity"?
And we could remove `never_loggedin` for now.
What do others think?

**Comment by Frenzie** [comment]
Sure, last login could be weeks or months  ago unless it actually  means
last activity.

**Review by equinoxmatt — COMMENTED** [review]
> Not a strong opinion, but maybe instead of login we could say a more general "last user activity"?
> And we could remove `never_loggedin` for now.
> What do others think?

'Last seen' maybe a good option?

File: app/Models/Auth.php [hunk 1]
```diff
@@ -105,6 +105,8 @@ public static function giveAccess() {
 
 		Minz_Session::_param('loginOk', self::$login_ok);
 		Minz_Session::_param('REMOTE_USER', httpAuthUser());
+		FreshRSS_UserDAO::touch($current_user);
```
[the remark this answers is not in the document]
Comment (Alkarex) [reply app/Models/Auth.php]: Actually, I had forgotten but we already have this line higher up, in the `init()` function

**Comment by Alkarex** [comment]
It depends a bit on what we want to measure. In my opinion, the most interesting are user actions, which are typically not automatised.

Here are some examples (it should already now be more or less like that, but some points need to be double-checked):

Catching:
* Explicit Web login (and only explicit)
* Mark as read, favourite, label (from Web or API)
* Add/remove feed/category (from Web or API)
* Change settings
* Add/remove user (for admins)

Not catching:
* API synchronisation (often automatic)
* Viewing FreshRSS RSS outputs (typically automatic)
* Anonymous views
* Pull-based synchronisation by cron
* WebSub updates
* Automatic database cleaning

**Comment by equinoxmatt** [comment]
I am not sure if this is getting overcomplicated? The original issue was to fix a use case whereby an admin can quickly see which of their users are using the system.  Last login is an ideal way to measure that. Even if they logged in and then straight back out, they are still using the system.

This is why my initial approach was recording the login time as opposed to when the configuration file was last modified.

I have also just discovered, when the actualize script runs, all the configuration files are touched. Therefore, you *cannot* rely on the configuration file modified time. Background processes modify it.

I propose we either put the last_login value back in the user config, as per my original PR or leave this issue until we have a user table.

**Comment by Alkarex** [comment]
My last post was only to explain the choice of the wording, not to ask for more work :-P
For instance, you might have an active user, using solely the API (e.g. from a mobile app), and never (or rarely) logging in. Therefore the login date is not a robust indication.

**Comment by equinoxmatt** [comment]
> My last post was only to explain the choice of the wording, not to ask for more work :-P
> For instance, you might have an active user, using solely the API (e.g. from a mobile app), and never (or rarely) logging in. Therefore the login date is not a robust indication.

I see your point. How do you want to proceed with this? I don't think the file modified time can realistically be used. As I said previously, it can be modified by background processes.

**Comment by Alkarex** [comment]
I think we should complete this PR with the file modified time (which should still work OK even if not perfect), and wait for another PR to improve the storage of this information in a dedicated field in DB

File: app/views/user/manage.phtml [hunk 2]
```diff
@@ -77,20 +77,20 @@
...
 				<tr>
-					<td><?= $username ?></td>
+					<td><a href="<?= _url('user', 'details', 'username', $username) ?>"><?= $username ?></a></td>
 					<td><?= $values['is_admin'] ? '✔' : ' ' ?></td>
 					<td><?= $values['mail_login'] ?></td>
 					<td><?= _t("gen.lang.{$values['language']}") ?></td>
 					<td><?= format_number($values['feed_count']) ?></td>
 					<td><?= format_number($values['article_count']) ?></td>
 					<td><?= format_bytes($values['database_size']) ?></td>
-					<td><a href="<?= _url('user', 'details', 'username', $username) ?>">Details</a></td>
```
Comment (Alkarex) [inline app/views/user/manage.phtml:93]: This line was not translated (from before this PR), so I moved the link

**Comment by Alkarex** [comment]
@equinoxmatt I have made a few changes. Could you please double-check whether that works for you?

**Comment by Alkarex** [comment]
Merging, but feedback still welcome

**Comment by equinoxmatt** [comment]
@Alkarex Sorry for the delay, the changes LGTM 👍  thansk for your help on this.

**Comment by alexkuang0** [comment]
Does this `last user activity` field updates when users synchronize their feed using RSS clients only?

**Comment by Alkarex** [comment]
@alexkuang0 The last user activity is updated only when there is a non-automatic action such as marking an article as read, as favourite, modifying a feed / category, etc. independently of whether this is done through the Web interface or through the API.
Login in through the Web interface is also counted as an activity.

**Pull request merged**